src/query/storage/promremote/storage.go

```diff
@@ -164,8 +164,14 @@ func (p *promStorage) writeSingle(
 			p.logger.Error("error reading body", zap.Error(err))
 			response = errorReadingBody
 		}
-		return fmt.Errorf("expected status code 2XX: actual=%v, address=%v, resp=%s",
-			resp.StatusCode, address, response)
+		genericError := fmt.Errorf(
+			"expected status code 2XX: actual=%v, address=%v, resp=%s",
+			resp.StatusCode, address, response,
+		)
+		if resp.StatusCode < 500 && resp.StatusCode != http.StatusTooManyRequests {
+			return xerrors.NewInvalidParamsError(genericError)
+		}
```

Comment:
There actually is an error type which carries the response code. Though I'm not exactly sure if somewhere else in the stack it won't get wrapped in some other error and as such lose the status code.

https://github.com/m3db/m3/blob/bd4237c36cfaca6cdf6620088b1eadbc885274b9/src/x/net/http/errors.go#L60-L62

Also, should probably discuss with others if this error should mostly be used API/http layers of the codebase, or if it's ok to use it deeper.

Reply:
I don't think it's a good idea to throw it from storage layer. It seems currently its only thrown form handlers and middleware which are HTTP related things.

Reply:
Also current error handling in HTTP layer is unique for this handler https://github.com/m3db/m3/blob/013972931d01f3019a7089680e8fa2d135bc9424/src/query/api/v1/handler/prometheus/remote/write.go#L321

So I just played it safe to not affect other storages.